lighthouse-cli/run.js

```diff
@@ -159,66 +163,62 @@ async function saveResults(runnerResult, flags) {
+ * @param {ChromeLauncher.LaunchedChrome=} launchedChrome
+ * @return {Promise<void>}
+ */
+async function potentiallyKillChrome(launchedChrome) {
+  if (!launchedChrome) return;
+
+  try {
+    await launchedChrome.kill();
+  } catch (err) {
+    process.stderr.write(`Couldn't quit Chrome process. ${err.toString()}\n`);
+  }
+}
+
 /**
  * @param {string} url
  * @param {LH.CliFlags} flags
  * @param {LH.Config.Json|undefined} config
- * @return {Promise<LH.RunnerResult|void>}
+ * @return {Promise<LH.RunnerResult|undefined>}
  */
-function runLighthouse(url, flags, config) {
-  /** @type {ChromeLauncher.LaunchedChrome|undefined} */
-  let launchedChrome;
-  const shouldGather = flags.gatherMode || flags.gatherMode === flags.auditMode;
-  let chromeP = Promise.resolve();
-
-  process.on('unhandledRejection', async (reason) => {
+async function runLighthouse(url, flags, config) {
+  /** @param {any} reason */
+  async function handleTheUnhandled(reason) {
     process.stderr.write(`Unhandled Rejection. Reason: ${reason}\n`);
-    try {
-      await potentiallyKillChrome();
-    } catch (err) {
-      process.stderr.write(`Couldn't quit Chrome process. ${err.toString()}\n`);
-    }
+    await potentiallyKillChrome(launchedChrome);
     setTimeout(_ => {
       process.exit(1);
     }, 100);
-  });
-
-  if (shouldGather) {
-    chromeP = chromeP.then(_ =>
-      getDebuggableChrome(flags).then(launchedChromeInstance => {
-        launchedChrome = launchedChromeInstance;
-        flags.port = launchedChrome.port;
-      })
-    );
   }
+  process.on('unhandledRejection', handleTheUnhandled);
 
-  const resultsP = chromeP.then(_ => {
-    return lighthouse(url, flags, config).then(runnerResult => {
-      return potentiallyKillChrome().then(_ => runnerResult);
-    }).then(async runnerResult => {
-      // If in gatherMode only, there will be no runnerResult.
-      if (runnerResult) {
-        await saveResults(runnerResult, flags);
-      }
+  /** @type {ChromeLauncher.LaunchedChrome|undefined} */
+  let launchedChrome;
 
-      return runnerResult;
-    });
-  });
+  try {
+    const shouldGather = flags.gatherMode || flags.gatherMode === flags.auditMode;
+    if (shouldGather) {
+      launchedChrome = await getDebuggableChrome(flags);
+      flags.port = launchedChrome.port;
+    }
 
-  return resultsP.catch(err => {
-    return Promise.resolve()
-      .then(_ => potentiallyKillChrome())
-      .then(_ => handleError(err));
-  });
+    const runnerResult = await lighthouse(url, flags, config);
 
-  /**
-   * @return {Promise<{}>}
-   */
-  function potentiallyKillChrome() {
-    if (launchedChrome !== undefined) {
-      return launchedChrome.kill();
+    // If in gatherMode only, there will be no runnerResult.
+    if (runnerResult) {
+      await saveResults(runnerResult, flags);
     }
-    return Promise.resolve({});
+
+    await potentiallyKillChrome(launchedChrome);
+    process.removeListener('unhandledRejection', handleTheUnhandled);
```

Comment:
should this be removed in `catch` too?

Reply:
> should this be removed in `catch` too?

`handleError` always ends with a `process.exit()`, so it doesn't matter in the `catch` case. If we moved `handleError` to `bin.js`, though, we could put both `potentiallyKillChrome` and removing the listener to a `finally` block, which would be kind of nice.

Reply:
oh right 👍